src/urllib3/response.py

```diff
@@ -986,6 +986,9 @@ def read(
 
         if amt is None:
             data = self._decode(data, decode_content, flush_decoder)
+            if len(self._decoded_buffer) != 0:
+                data = self._decoded_buffer.get_all() + data
```

Comment:
Let's rely on joining capability of the buffer which may be more efficient, and have a small optimization for the case when `data` is empty

```suggestion
                if data:
                    self._decoded_buffer.put(data)
                data = self._decoded_buffer.get_all()
```